File: src/BjjEire.Domain/Entities/Competitions/Competition.cs
// Copyright (c) BjjEire. All rights reserved.
// Licensed under the MIT License.

namespace BjjEire.Domain.Entities.Competitions;


public class Competition : BaseEntity
{
    [BsonElement("slug")]
    public string Slug { get; set; } = string.Empty;

    [BsonElement("name")]
    public string Name { get; set; } = string.Empty;

    [BsonElement("description")]
    public string? Description { get; set; }

    [BsonElement("organisation")]
    public string Organisation { get; set; } = string.Empty;

    [BsonElement("country")]
    public string Country { get; set; } = "Ireland";

    [BsonElement("websiteUrl")]
    public string WebsiteUrl { get; set; } = string.Empty;

    [BsonElement("registrationUrl")]
    public string? RegistrationUrl { get; set; }

    [BsonElement("logoUrl")]
    public string? LogoUrl { get; set; }

    [BsonElement("tags")]
    public List<string> Tags { get; set; } = [];

    [BsonElement("startDate")]
    public DateTime? StartDate { get; set; }

    [BsonElement("endDate")]
    public DateTime? EndDate { get; set; }

    [BsonElement("isActive")]
    public bool IsActive { get; set; } = true;
}

File: src/bjjeire-api/src/main/java/com/bjjeire/api/competition/Competition.java
package com.bjjeire.api.competition;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(Competition.ENTITY_NAME)
public class Competition {

    public static final String ENTITY_NAME = "Competition";
    public static final Duration EXPIRY_GRACE = Duration.ofDays(365L * 2);

    @Id
    private String id;

    @Field("expiresAt")
    private Instant expiresAt;

    private String slug;
    private String name;
    private String description;
    private String organisation;

    @Builder.Default
    private String country = "Ireland";

    private String websiteUrl;
    private String registrationUrl;
    private String logoUrl;

    @Builder.Default
    private List<String> tags = new ArrayList<>();

    private Instant startDate;
    private Instant endDate;

    @Builder.Default
    private boolean isActive = true;

    @Field("createdAt")
    private Instant createdOnUtc;

    @Field("createdBy")
    private String createdBy;

    @Field("updatedAt")
    private Instant updatedOnUtc;

    @Field("updatedBy")
    private String updatedBy;

    public Instant computeExpiresAt() {
        return endDate != null ? endDate.plus(EXPIRY_GRACE) : null;
    }

    public void stampExpiry() {
        expiresAt = computeExpiresAt();
    }
}
